weave-js/src/components/PagePanelComponents/Home/Browse3/pages/ObjectVersionsPage.tsx

```diff
@@ -154,65 +172,126 @@ const ObjectVersionsTable: React.FC<{
-            },
-          }),
-        ]
-      : []),
-  ];
-  const columnGroupingModel: GridColumnGroupingModel = [];
+
+  // TODO: We should make this page very robust similar to the CallsTable page.
+  // We will want to do nearly all the same things: URL state management,
+  // sorting, filtering, ref expansion, etc... A lot of common logic should be
+  // extracted and shared.
+  const {cols: columns, groups: columnGroupingModel} = useMemo(() => {
+    let groups: GridColumnGroupingModel = [];
+    const cols: GridColDef[] = [
+      basicField('object', 'Object', {
+        hideable: false,
+        renderCell: cellParams => {
+          // Icon to indicate navigation to the object version
+          const obj: ObjectVersionSchema = cellParams.row.obj;
+          return (
+            <ObjectVersionLink
+              entityName={obj.entity}
+              projectName={obj.project}
+              objectName={obj.objectId}
+              version={obj.versionHash}
+              versionIndex={obj.versionIndex}
+              fullWidth={true}
+            />
+          );
+        },
+      }),
+    ];
+
+    cols.push(
+      basicField('baseObjectClass', 'Category', {
+        width: 100,
+        valueGetter: cellParams => {
+          return cellParams.row.obj.baseObjectClass;
+        },
+        renderCell: cellParams => {
+          const category = cellParams.value;
+          if (category === 'Model' || category === 'Dataset') {
+            return <TypeVersionCategoryChip baseObjectClass={category} />;
+          }
+          return null;
+        },
+      })
+    );
+
+    if (!props.usingLatestFilter) {
+      const dynamicFields: string[] = [];
+      const dynamicFieldSet = new Set<string>();
+      rows.forEach(r => {
+        Object.keys(r.obj.val).forEach(k => {
+          if (!dynamicFieldSet.has(k)) {
+            dynamicFieldSet.add(k);
+            dynamicFields.push(k);
+          }
+        });
+      });
+
+      const {cols: newCols, groupingModel} =
+        buildDynamicColumns<ObjectVersionSchema>(dynamicFields, (row, key) => {
+          const obj: ObjectVersionSchema = (row as any).obj;
+          const res = obj.val?.[key];
+          if (isTableRef(res)) {
+            // This whole block is a hack to make the table ref clickable. This
+            // is The same thing that the CallsTable does for expanded fields.
+            // Once we come up with a common pattern for ref expansion, this
+            // will go away.
+            const selfRefUri = objectVersionKeyToRefUri(obj);
+            const targetRefUri =
+              selfRefUri +
+              ('/' +
+                OBJECT_ATTR_EDGE_NAME +
+                '/' +
+                key.split('.').join(OBJECT_ATTR_EDGE_NAME + '/'));
+            return makeRefExpandedPayload(targetRefUri, res);
+          }
+          return res;
+        });
+      cols.push(...newCols);
+      groups = groupingModel;
+    }
+
+    cols.push(
```

Comment:
Perhaps for discussion with design, I know we put the equivalent column on the right for calls but I feel like it would be better to position this column to the left of the dynamic fields. When viewing the versions of an object that feels like important information. (Better longterm solution is to allow users to set their own column ordering and preserve that in state.)

Reply:
totally agree - I think this page will likely go through a few rounds of such revisions